servant-server/src/Servant/Server/Internal.hs

```diff
@@ -703,3 +707,54 @@ instance (HasContextEntry context (NamedContext name subContext), HasServer subA
       subContext = descendIntoNamedContext (Proxy :: Proxy name) context
 
   hoistServerWithContext _ _ nt s = hoistServerWithContext (Proxy :: Proxy subApi) (Proxy :: Proxy subContext) nt s
+
+-------------------------------------------------------------------------------
+-- TypeError helpers
+-------------------------------------------------------------------------------
+
+#if MIN_VERSION_base(4,9,0)
+-- | This instance catches mistakes when there are non-saturated
+-- type applications on LHS of ':>'.
+--
+-- >>> serve (Proxy :: Proxy (Capture "foo" :> Get '[JSON] Int)) (error "...")
+-- ...
+-- ...Expected something of kind Symbol or *, got: k -> l on the LHS of ':>'.
+-- ...Maybe you haven't applied enough arguments to
+-- ...Capture "foo"
+-- ...
+instance TypeError
+    ('Text "Expected something of kind Symbol or *, got: k -> l on the LHS of ':>'."
+    ':$$: 'Text "Maybe you haven't applied enough arguments to"
+    ':$$: 'ShowType arr)
+    => HasServer ((arr :: k -> l) :> api) context
+  where
+    type ServerT (arr :> api) m = Void
+    -- it doens't really matter what sub route we peak
+    route Proxy context d = route (Proxy :: Proxy Raw) context (vacuous d)
+    hoistServerWithContext _ _ _ = id
+
+-- | This instance prevents from accidentally using '->' instead of ':>'
+--
+-- >>> serve (Proxy :: Proxy (Capture "foo" Int -> Get '[JSON] Int)) (error "...")
+-- ...
+-- ...No instance HasServer (a -> b).
+-- ...Maybe you have used '->' instead of ':>' between
+-- ...Capture "foo" Int
+-- ...and
+-- ...Verb 'GET 200 '[JSON] Int
+-- ...
+instance TypeError
+    ('Text "No instance HasServer (a -> b)."
```

Comment:
Don't you want `'ShowType a :<>: 'Text "->" :<>: 'ShowType b` instead of `a -> b`?

Reply:
No I don't want. The `a` and `b` are printed below separately (as they can be huge).

Reply:
I fear that `a` and `b` in `a -> b` might confuse users who might think they have some ambiguous type variables or maybe they'll actually have `a` or `b` in their program.

OTOH I agree that huge types are not better.